rmi: don't break when the image is missing a manifest

nalind opened the pull request:
In `libpod/image/Image.Remove()`, don't go looking for the image's parent before actually deleting the image.  If the attempt to find the image's parent fails for any reason, we're basically stuck, when we could fail in a way that at least gets the caller closer to where they wanted to be.

In `libpod/Runtime.RemoveImage()`, if we can't determine if an image has children because of some error, assume that it doesn't instead of bailing out.

Combined, these changes allow removal of an image that's lost its manifest for whatever reason, as we're seeing in https://bugzilla.redhat.com/show_bug.cgi?id=1884043.

mheon commented:
Tests are red, but changes LGTM
/approve
@vrothberg PTAL

nalind commented:
I missed that we're not able to correctly calculate an image's "parent" after we delete it, which is what's breaking the tests.
/retitle WIP: rmi: don't break when the image is missing a manifest

rhatdan commented:
Can't we just make the parent not existing and expected error and just log it was debugf or warnf.  And continue on?

rhatdan commented on libpod/image/image.go:
```diff
@@ -615,14 +615,14 @@ func (i *Image) TopLayer() string {
 func (i *Image) Remove(ctx context.Context, force bool) error {
-	parent, err := i.GetParent(ctx)
-	if err != nil {
-		return err
```
If you just log.Warnf(err) here, everything else will work fine, I believe.

nalind replied:
Tried it, and it still failed with another file-not-found I haven't chased down yet.  It's mixed in with another error, so trace logging still only gives me a backtrace that points to `pkg/errorhandling.JoinErrors()`.  Frustrating, that.

nalind replied:
Never mind, `rmi -a` apparently doesn't use the same code path as `rmi`.

nalind commented:
/retitle rmi: don't break when the image is missing a manifest

nalind commented:
Updated tests to not try `podman system reset` when testing the remote client, which doesn't implement `system reset`, and added tests that try removing the config blob instead of the manifest.

rhatdan commented on test/system/010-images.bats:
```diff
@@ -241,4 +241,94 @@ Labels.created_at | 20[0-9-]\\\+T[0-9:]\\\+Z
+	rm -v ${graphroot}/*-images/$id/=$(echo -n "sha256:$id" | base64 -w0)
+	run_podman 125 images
+	run_podman rmi -f $id
+	if [[ "$PODBIN_NAME" != remote ]] ; then
```
Why do you need to reset storage?  And if you don't reset it on remote, why isn't it broken then?

nalind replied:
Hmm, yeah, that should probably just be a different test.

rhatdan commented:
@edsantiago PTAL at the tests.

edsantiago commented:
Looking. I don't like the duplication, will offer a suggestion later. And I *REALLY* don't like the `system reset`... but don't see a way around it.

First things first though: the final test barfs (root) with:
```
...
   [ rc=125 (expected) ]
   # /home/esm/src/atomic/2018-02.podman/libpod/bin/podman system reset -f
   time="2021-04-19T14:45:02-06:00" level=warning msg="error determining parent of image: error reading image \"961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4\" as image: error locating item named \"manifest\" for image with ID \"961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4\": file does not exist, ignoring the error"
   A storage.conf file exists at /etc/containers/storage.conf
   You should remove this file if you did not modified the configuration.
   time="2021-04-19T14:45:02-06:00" level=error msg="unlinkat /var/lib/containers/storage/overlay-containers/1d9ac53d745d3ed1ce5d981b072ee002f52d23b7a8a0471f376aa7044044a0b0/userdata/shm: device or resource busy"
   [ rc=125 (** EXPECTED 0 **) ]
   #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
   #| FAIL: exit code is 125; expected 0
   #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

nalind commented:
Why is there a container present?  The test doesn't create one.

edsantiago commented:
There are no containers: `images -a` and `ps -a` are empty. There was a stray mount, probably a leftover from some prior test run. `umount /var/lib/containers/.../shm` makes the failure go away.

edsantiago commented:
Tests need a pretty major rewrite; I'm going to have to finish it tomorrow, sorry. (Quick rundown: too much duplication, and we need a `teardown()` because any failure anywhere in a test will leave the entire system unusable. This is podman, where tests use `/var/lib/containers`, not buildah which uses a custom root for each test.)

edsantiago commented:
Help, please. To minimize `podman pull`s, I'm trying `podman save/load`, and that completely ruins everything. Basically, if I `podman load` a saved image, and `rm .../manifest`, `podman images` still complains but now it exits 0 (success) instead of 125:
```bash
#!/bin/bash

# SETUP
img=quay.io/libpod/alpine@sha256:634a8f35b5f16dcf4aaa0822adc0b1964bb786fca12f6831de8ddc45e5986a00
id=961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4

tmpdir=$(mktemp -d --tmpdir podman_manifest_test.XXXXXXX)
mkdir $tmpdir/root

PODMAN="./bin/podman --root $tmpdir/root"

$PODMAN pull -q $img
$PODMAN save -o $tmpdir/img.tar $img

# Test 1
echo
echo "Test 1: this 'podman images' will fail"
rm -v $tmpdir/root/*-images/$id/manifest
$PODMAN images
echo status=$? expect=125

umount $tmpdir/root/overlay

echo "--------------------------------------------------------------"
rm -rf $tmpdir/root
mkdir $tmpdir/root
$PODMAN load -q -i $tmpdir/img.tar

echo
echo "Test 2: this 'podman images' will succeed????"
rm -v $tmpdir/root/*-images/$id/manifest
$PODMAN images
echo status=$? expect=125
```
Is this the same as #3761? Is it even a bug? Whatever it is: how can I work around this?

nalind commented:
Apparently we weren't treating this as an error for images with no names attached to them.  Manually untagging the image after pulling it also sidesteps the error.  `images -a` still produces an error when it fails to retrieve labels from the image's configuration blob, which we can't find without consulting the manifest.

nalind commented:
Conversely, manually tagging the image after re-loading it but before removing its manifest will re-trigger the errors.

edsantiago commented:
That works - thank you!

edsantiago commented:
@nalind would you consider this instead? (Save as [330-corrupt-images.bats](https://github.com/containers/podman/files/6345333/330-corrupt-images.bats.txt))

I choose to disable entirely for podman-remote, because there's no sane way to restart the server for each test. If the code paths in podman-remote are sufficiently different that they MUST be tested, lmk, I can (grumblingly) make this work with the default root.

nalind commented:
I'm not particularly invested in testing with the remote-client configuration (Red Hat bug 1884043 is concerned with local podman on an OpenShift node), but the CI configuration was running the ones I was adding with it.  Removed the tests I was adding in favor of yours, rebasing, and repushing.

nalind commented:
Rebased.

edsantiago commented on test/system/330-corrupt-images.bats:
```diff
@@ -0,0 +1,134 @@
+}
+
+@test "podman corrupt images - system reset" {
+    _corrupt_image_test "image prune -a -f"
```
Oops. The nitwit you copied this from (who prefers to remain anonymous right now) seems to have forgotten to actually fix this line. Sorry.

edsantiago commented:
/lgtm
/hold

The aforementioned nitwit has privately told me that he will volunteer to clean up the 'prune' in a followup PR.

vrothberg reviewed (commented):
/lgtm

vrothberg commented:
/hold cancel

rhatdan commented:
/hold cancel
